internal/output/outputter.go

```diff
@@ -0,0 +1,147 @@
+	indentation           int
+	curBuildStep          int
+	curPipelineStep       int
+	pipelineStepDurations []time.Duration
+	curPipelineStart      time.Time
+	curPipelineStepStart  time.Time
+}
+
+func Get(ctx context.Context) *Outputter {
+	val := ctx.Value(outputterContextKey)
+
+	if val != nil {
+		return val.(*Outputter)
+	}
+
+	// No outputter found in context, something is wrong.
+	panic("Called output.Get(ctx) on a context with no Outputter attached!")
+}
+
+func NewOutputter(logger logger.Logger) Outputter {
+	return Outputter{logger: logger}
+}
+
+func WithOutputter(ctx context.Context, outputter Outputter) context.Context {
+	return context.WithValue(ctx, outputterContextKey, &outputter)
+}
+
+func (o *Outputter) printColorf(color Color, format string, a ...interface{}) {
+	o.logger.Infof("%s%s%s", string(color), fmt.Sprintf(format, a...), cReset)
+}
+
+func (o *Outputter) StartPipeline() {
+	o.printColorf(cBlue, "──┤ Pipeline Starting … ├────────────────────────────────────────")
+	o.curPipelineStep = 1
+	o.pipelineStepDurations = nil
+	o.curPipelineStart = time.Now()
+}
+
+func (o *Outputter) EndPipeline() {
+	for i, duration := range o.pipelineStepDurations {
+		o.logger.Infof("  │ Step %d - %.3fs", i+1, duration.Seconds())
+	}
+
+	elapsed := time.Now().Sub(o.curPipelineStart)
+	o.logger.Infof("%s──┤ ︎Pipeline Done in %s%.3fs%s ⚡︎├────────────────────────────────────%s",
+		cBlue,
+		cGreen,
+		elapsed.Seconds(),
+		cBlue,
+		cReset)
+	o.curPipelineStep = 0
+}
+
+// how many steps there are per pipeline
+// we might need to change or remove this if it turns out we can't predict how many steps there will be
+const numSteps = 2
+
+func (o *Outputter) StartPipelineStep(format string, a ...interface{}) {
+	o.printColorf(cGreen, "STEP %d/%d — %s", o.curPipelineStep, numSteps, fmt.Sprintf(format, a...))
+	o.curPipelineStep++
+	o.curBuildStep = 1
+	o.curPipelineStepStart = time.Now()
+}
+
+func (o *Outputter) EndPipelineStep() {
+	elapsed := time.Now().Sub(o.curPipelineStepStart)
+	o.logger.Infof("    (Done %.3fs)", elapsed.Seconds())
+	o.logger.Infof("")
+	o.pipelineStepDurations = append(o.pipelineStepDurations, elapsed)
+}
+
+func (o *Outputter) StartBuildStep(format string, a ...interface{}) {
+	o.logger.Infof("  → %s", fmt.Sprintf(format, a...))
+	o.curBuildStep++
+}
+
+func (o *Outputter) Print(format string, a ...interface{}) {
+	if o.curBuildStep == 0 {
+		o.logger.Infof(format, a)
+	} else {
+		o.logger.Infof("%s%s", buildStepOutputPrefix, fmt.Sprintf(format, a...))
+	}
+}
+
+// sticks "prefix" at the start of every new line
+type prefixedWriter struct {
+	prefix                string
+	underlying            io.Writer
+	indentBeforeNextWrite bool
+}
+
+var _ io.Writer = &prefixedWriter{}
+
+func newPrefixedWriter(prefix string, underlying io.Writer) *prefixedWriter {
+	return &prefixedWriter{prefix, underlying, true}
+}
+
+func (i *prefixedWriter) Write(buf []byte) (n int, err error) {
+	for _, c := range buf {
```

Comment:
hmmm...writing one character at a time isn't really how the Writer api is supposed to be used, and i'm a little worried this will lead to weird behavior (e.g., defeating buffering on the underlying Writer, breaking multi-byte chars like emoji)

have you considered rewriting the bytes up-front, with something like
s := bytes.Replace(buf, '\n', '\n' + prefix, 0)
i.underlying.Write(s)
?

(obviously, with some extra logic to handle the edge cases)

Reply:
I suspect it would defeat buffering, but I wouldn't expect it to break multi-byte chars (at least, my experience with unix files is that the way in which you break up the bytes over write calls isn't supposed to affect what comes out).

I sillily figured the single byte implementation would be simpler, but that only really makes sense in terms of allocations, which is pretty unlikely to matter for us, so yeah, might as well go with string replacement